### PR TITLE
chore(compactor): move CompactRange up to RoundInfo

### PR DESCRIPTION
**Pull request by Joe-Blount**
**What's changing (short description)**
CompactRanges is gone as a RoundInfo type (i.e. its no longer a member of the enum with ManySmallFiles, TargetLevel, etc). Now CompactRange is its own data type. The RoundInfo has a vec of CompactRanges, each of which describes a time range of the partition and has an opType (operation type), which is ManySmallFiles, TargetLevel, etc.
The CompactRanges have their `own files_for_now`, `files_for_later`, etc.  Each CompactRange is operated on as if its the entire world.  For example, some CompactRanges might be doing ManySmallFiles while others are doing TargetLevel.

**Why its changing (value-add)**
Because L0s can overlap, and customers can "spray the data" with random backfilling, we can have very complicated situations to unwind with overlapped L0s.  The highly backlogged situations (before this PR) don't necessarily behave the same as the moderately sized test cases we have.  I think efficiency drops in highly backlogged situations (10k l0s, 50k, 100k, etc).
With this PR, we split highly backlogged situations into multiple CompactRanges, then have a firmly bounded problem size as we compact the L0s to L1.  The CompactRange aims for around `max_compact_size` of L0s (regardless of L0 file count).
That means whether we have enough backlog for 3 CompactRegions or 300, the backlog is split up into manageable CompactRanges and the compaction decisions are made on this consistently sized, predictably behaving set of files.  So linear increases in backlog should result in linear increases in compaction work to catch up.
It also makes it far easier (for me, at least) to reason about the decisions made in Compaction.  If there's hundreds of "runs" in a simulation, I find it difficult to absorb the big picture and form opinions on whether or not its making good decisions.  With CompactRanges, I can follow the time range through the runs and consider just the decisions on a consistently scoped range of files.

We could probably improve the simulator to somehow obtain knowledge of the CompactRanges, and display it, making it easier to follow CompactRanges through their life cycle in simulation output.  But that's an idea for another PR.

L1->L2 compactions aren't nearly the challenge that L0->L1 compactions are, and I don't want to artificially split L2s at the CompactRange boundaries.  So for now, when we have multiple CompactRanges, all the L2s are in the CompactRange's `files_for_later` (because if we tried to put them in CompactRanges, they'd span the range boundaries, which means CompactRanges couldn't be operated on without regard for other ranges).
There are options to change this in the future, if desired.  We could conditionally put L2s in the CompactRanges (when they don't span them).  Or just put them in the first (left most) CompactRange, etc.  There's pros & cons to the various options, but these are options for a later PR.  Or it could stay like it is.

**Cleanup in this PR:**
- Some of the ManySmallFiles complications where it was trying to infer past decisions is removed in this PR.  Its not moved, that complexity is now obsolete by the better control provided by RoundInfo & CompactRanges.  So its just deleted.  When I deleted it, one layout test changed write amplification by 0.1gb.  
- On a previous PR there was a request to split up the functionality in `vertical_split_handling` (it both decided to split and took the opportunity to identify previous splits).  Detecting previous splits (to produce CompactRanges) is now done well before the decision to do additional splitting.  So this functionality is now split up and moved around.

**Write Amplification impact of this PR**
Write amplification impacts of this PR are mixed.  Some get better, some get worse.  Overall, :shrug:, this cleans us up some and gives better control over the decisions without significantly impacting write amp, so that's a win from my perspective.  In the interest of not growing this too big, I quit looking at write amplification improvements when I got it to be comparable overall.

Summary of write amplification changes:
random_backfill_empty_partition  1.39gb -> 1.8gb
random_backfill_over_l2s 3.35gb -> 3.15gb 
actual_case_from_catalog_1 15.63gb -> 13.84gb
many_l0_files_different_created_order_non_overlap 25kb -> 18kb (its an L1 output instead of L2, but it's still all compacted to single file, so equivalent)
not_many_l0_and_overlapped_l1_files 21mb -> 28mb
l0s_needing_vertical_split 2.62gb -> 2.63gb
many_medium_files_time_range_1 1.46gb -> 1.17 gb
stuck_l0 26.51gb -> 27.19gb
split_precent_loop 1.98gb -> 3.12gb
two_giant_files_time_range_1 4.88gb (skipped) -> 9.77gb  (previously the partition was skipped, so this is better)

**Potential near term follow on PRs / cleanup for later:**
 - Cap the number of active CompactRegions.  Assume there's 50k L0 files and based on their size they should be split into 500 CompactRegions.  Instead of trying to work on all 500 regions at once, we're probably better off to pick a number that gives good concurrency (low 10s?), and only work on that many, starting at the left (oldest).  This should avoid the big surges in file counts I've sometimes seen on highly backlogged partitions, and give a more steady decline in L0 count.
- Currently CompactType::Deferred is overused for regions that are deferred and regions we haven't decided what partition to perform on.  We could introduce a new op type for unknown/tbd/etc so deferred more cleanly means deferred.
- After the number of active CompactRegions is capped, it will be time to delete SimulatedLeadingEdge.  The temporary motivation for that was to have a brute force incremental way to deal with many L0 files that confuse the standard algorithms.  With large backlogs getting split into many CompactRegions (and once theres a limit on how many CompactRegions we consider before deferring the rest), the motivation for this workaround will be removed.
- As mentioned above, keeping L2s out of CompactRanges could change.

helps: https://github.com/influxdata/idpe/issues/17974

- [x] I've read the contributing section of the project [README](https://github.com/influxdata/influxdb_iox/blob/main/README.md).
- [x] Signed [CLA](https://influxdata.com/community/cla/) (if not already signed).

### CODE REVIEW

File: compactor/src/components/round_info_source/mod.rs [hunk 1]
```diff
@@ -208,197 +211,425 @@ impl LevelBasedRoundInfo {
...
+
+        // As we iterate through the last_round_info's ranges, we'll try to consolidate ranges for any that don't have L0s.
+        let mut prior_range: Option<CompactRange> = None;
+
+        for range in &last_round_info.ranges {
+            // The prior pround should have handled its `files_for_now`, so that should `None`.
+            // What the prior round considered `files_for_later` will now become `files_for_now`.
+            assert!(
+                range.files_for_now.lock().unwrap().is_none(),
+                "files_for_now should be empty for range {}->{} on partition {}",
```
Comment (wiedld) [inline compactor/src/components/round_info_source/mod.rs:374]: Important: These assertions are the only way that we can be sure of the desired outcome. So we cannot remove, and retain code confidence.

But then can we please, please watch this carefully on the hours after it merges. Because this is how we got compactor panics. 🙏🏼

File: compactor/src/components/round_info_source/mod.rs [hunk 2]
```diff
@@ -208,197 +211,425 @@ impl LevelBasedRoundInfo {
...
-                    .sum::<usize>();
-                let cap = l0cap
-                    + overlaps
+                let mut split_ranges = Vec::with_capacity(split_times.len());
+                let mut this_split: Vec<ParquetFile>;
+                let mut max = range.max;
+
+                for split_time in split_times.into_iter().rev() {
+                    // By iterating in reverse, everything above the split time is in this split
+                    (this_split, files_for_now) = files_for_now
```
Comment (wiedld) [inline compactor/src/components/round_info_source/mod.rs:418]: Please rename `(this_split, files_for_later)` because the latter partitioned files are not for now.

In fact, that variable `files_for_now` is shadowed later as  = `Some(this_split)`.

File: compactor/src/components/round_info_source/mod.rs [hunk 3]
```diff
@@ -208,197 +211,425 @@ impl LevelBasedRoundInfo {
...
+                }
 
-                if !overlaps.is_empty() {
-                    prior_overlapping_max = overlaps.iter().map(|f| f.max_time).max().unwrap();
-                    let prior_smallest_max = overlaps.iter().map(|f| f.max_time).min().unwrap();
-                    if prior_smallest_max < min {
-                        // Expand the range to include this file, so it can be included (if its small and we'd like to grow it).
-                        min = prior_smallest_max;
-                    }
+                if !files_for_now.is_empty() {
```
Comment (wiedld) [inline compactor/src/components/round_info_source/mod.rs:449]: With the variable shadowing, had to read through and determine which of the previous inner loops were `let files_for_now` (locally scoped), versus the outer loop `mut files_for_now`. Can we maybe reduce the variable shadow when it's not re-assignment?

File: compactor/src/components/round_info_source/mod.rs [hunk 4]
```diff
@@ -208,197 +211,425 @@ impl LevelBasedRoundInfo {
...
+            );
+            assert!(
+                range.branches.lock().unwrap().is_none(),
+                "branches should be empty for range {}->{} on partition {}",
+                range.min,
+                range.max,
+                partition_info.partition_id()
+            );
+
+            let files_for_now = range.files_for_later.lock().unwrap().take();
```
Comment (wiedld) [inline compactor/src/components/round_info_source/mod.rs:387]: Why is this `range.files_for_later` labeled as `files_for_now`?

File: compactor/src/components/round_info_source/mod.rs [hunk 5]
```diff
@@ -208,197 +211,425 @@ impl LevelBasedRoundInfo {
...
+                    let files_for_now = if this_split.is_empty() {
+                        None
+                    } else {
+                        Some(this_split.clone())
+                    };
+
+                    split_ranges.insert(
+                        0,
+                        CompactRange {
+                            op: CompactType::Deferred {},
```
Comment (wiedld) [inline compactor/src/components/round_info_source/mod.rs:434]: Question: why do all previously split regions get deferred for this round?

File: compactor/src/components/round_info_source/mod.rs [hunk 6]
```diff
@@ -208,197 +211,425 @@ impl LevelBasedRoundInfo {
...
+                };
+            }
+        }
+        if prior_range.is_some() {
+            ranges.push(prior_range.unwrap());
+        }
+
+        // If we still have several ranges, L2s (if any) need to stay in round_info.files_for_later.  If we have 1 range
+        // without L0s, the L2s can go in that range.
+        let mut deferred_l2s = last_round_info.take_files_for_later();
```
Comment (wiedld) [inline compactor/src/components/round_info_source/mod.rs:516]: Wait, how do we know the file_for_later (from the previous round) were all L2s? Because the last round  had L0s to compact, and would push-for-later any other files (including L0s and L1s) which are above the size limit.

**Review by wiedld — COMMENTED** [review]
There is a lot of code changed in this PR. Additionally, there is a lot of variable naming conventions which make it even harder to follow. Can we clean those up a bit?

File: compactor/src/components/divide_initial/multiple_branches.rs [hunk 7]
```diff
@@ -44,129 +39,85 @@ impl DivideInitial for MultipleBranchesDivideInitial {
...
-                                // Compacting a branch of 1 won't help us reduce the L0 file count.  Put it on the ignore list.
-                                more_for_later.push(current_branch.pop().unwrap());
-                            } else if !current_branch.is_empty() {
-                                branches.push(current_branch);
-                            }
-                            current_branch = Vec::with_capacity(capacity);
-                            current_branch_size = 0;
+                // Since its ManySmallFiles, we know the files are L0s, and the total bytes is under our limit.
+                // We just need to split them up into branches for compaction.
+                // TODO: it would be nice to pick some good split times, store them in the ManySmallFiles op, and use them consistently across all the branches.  That should make the later round more efficient.
```
[the remark this answers is not in the document]
Comment (Joe-Blount) [reply compactor/src/components/divide_initial/multiple_branches.rs:54]: I mean pick some good split times for where to split the output of compactions performed on behalf of `ManySmallFiles`.  It is a compaction operation from our perspective, but the timestamps at which it splits the output are still referred to as split times.

File: compactor/src/round_info.rs [hunk 8]
```diff
@@ -79,71 +73,201 @@ impl Display for RoundInfo {
...
+}
+
 impl RoundInfo {
+    /// take_files_for_later takes the files_for_later from the RoundInfo.
+    pub fn take_files_for_later(&self) -> Option<Vec<ParquetFile>> {
+        let files_for_later = self.files_for_later.lock().unwrap().take();
+        files_for_later
+    }
+}
+impl CompactType {
```
[the remark this answers is not in the document]
Comment (Joe-Blount) [reply compactor/src/round_info.rs:206]: good catch

File: compactor/src/components/round_info_source/mod.rs (hunk 1 above)
Comment (Joe-Blount) [reply compactor/src/components/round_info_source/mod.rs:374]: That's the plan :+1:

File: compactor/src/components/round_info_source/mod.rs (hunk 2 above)
Comment (Joe-Blount) [reply compactor/src/components/round_info_source/mod.rs:418]: `now` and `later` equate to `this round` and `some future round`.
so `files_for_now` means files we're likely to be working on in this round.
and `files_for_later` means files we've decided not work on in this round.
I've tried to be consistent on that, and think I have been.

This function is evaluating prior ranges, and updating them as appropriate for ranges in this round.  So what in prior range was something for `later` is the set of files for `now`.  That's why a few lines above it `range.files_for_later` and puts it in `files_for_now`  A few lines above that is a comment trying to explain that:
```
            // The prior round should have handled its `files_for_now`, so that should `None`.
            // What the prior round considered `files_for_later` will now become `files_for_now`.
```

Shadowing `files_for_now` is misleading.  I'll rename `this_split` to be `this_split_files_for_now` and change the shadowing so it is the reassignment.  Lets see if that helps.

File: compactor/src/components/round_info_source/mod.rs (hunk 3 above)
Comment (Joe-Blount) [reply compactor/src/components/round_info_source/mod.rs:449]: yeah, see above response, I think my next push will make it clearer.

File: compactor/src/components/round_info_source/mod.rs (hunk 4 above)
Comment (Joe-Blount) [reply compactor/src/components/round_info_source/mod.rs:387]: This function is evaluating prior ranges (the ranges from the previous round) and converting them into ranges for the current round.  What the previous round considered `for_later` (for a future round) is what we have to look at in this current round.  

The comment at the top of this loop says:
```
            // The prior round should have handled its `files_for_now`, so that should `None`.
            // What the prior round considered `files_for_later` will now become `files_for_now`.
```

File: compactor/tests/layouts/single_timestamp.rs [hunk 9]
```diff
@@ -125,25 +125,34 @@ async fn two_giant_files_time_range_1() {
...
-    - "L0.2[100,101] 2ns 4.88gb |------------------------------------------L0.2------------------------------------------|"
-    - "L0.3[100,100] 1ns 2.44gb |L0.3|                                                                                    "
-    - "L0.4[101,101] 1ns 2.44gb                                                                                           |L0.4|"
-    - "WARNING: file L0.2[100,101] 2ns 4.88gb exceeds soft limit 100mb by more than 50%"
+    - "**** Final Output Files (9.77gb written)"
+    - "L0, all files 2.44gb                                                                                               "
+    - "L0.3[100,100] 1ns        |L0.3|                                                                                    "
+    - "L0.4[101,101] 1ns                                                                                                  |L0.4|"
+    - "L0.5[100,100] 2ns        |L0.5|                                                                                    "
+    - "L0.6[101,101] 2ns                                                                                                  |L0.6|"
```
Comment (wiedld) [inline compactor/tests/layouts/single_timestamp.rs:151]: remark: so this is the edge case with massively dense data, where it splits each file (2 giant inputs at 2ns time_range each) into half. 👍🏼

**Review by wiedld — COMMENTED** [review]
Some code cleanups requested, please.
Overall, this is a very nice refactor. Thank you.

File: compactor/src/components/divide_initial/multiple_branches.rs (hunk 7 above)
[the remark this answers is not in the document]
Comment (Joe-Blount) [reply compactor/src/components/divide_initial/multiple_branches.rs:54]: When compaction output is split at certain timestamps, they're called split times [all the way to the PlanIR::Split](https://github.com/influxdata/influxdb_iox/blob/80f8b55baa2480d3692f0a7f5f344e205f6ec4a9/compactor/src/components/ir_planner/planner_v1.rs#L251)

File: compactor/src/components/divide_initial/multiple_branches.rs [hunk 10]
```diff
@@ -44,129 +39,85 @@ impl DivideInitial for MultipleBranchesDivideInitial {
...
-                    // L0 files can be highly overlapping, requiring 'vertical splitting' (see high_l0_overlap_split).
-                    // Achieving `vertical splitting` requires we tweak the grouping here for two reasons:
-                    //  1) Allow the large highly overlapped groups of L0s to remain in a single branch, so they trigger the split
-                    //  2) Prevent the output of a prior split from being grouped together to undo the previous veritacal split.
-
-                    // Both of these objectives need to consider the L0s as a chains of overlapping files.  The chains are
-                    // each a set of L0s that overlap each other, but do not overlap the other chains.
-                    // Chains can be created based on min_time/max_time without regard for max_l0_created_at because there
-                    // are no overlaps between chains.
-                    let initial_chains = split_into_chains(start_level_files);
```
[the remark this answers is not in the document]
Comment (Joe-Blount) [reply compactor/src/components/divide_initial/multiple_branches.rs:83]: :+1: I've been thinking that for a while (all the way back to not being confident about where I added them initially), but waiting to see how the rest of the code comes together and see if these functions stick around, where their primary users are, etc.  At this point it looks like they will stick around.  I just haven't moved them yet.

File: compactor/src/components/round_info_source/mod.rs [hunk 11]
```diff
@@ -409,152 +640,142 @@ impl RoundInfoSource for LevelBasedRoundInfo {
...
+                    .apply(partition_info, &files_for_now)
+                    .await?
+            {
+                return Ok((
+                    Arc::new(RoundInfo {
+                        ranges,
+                        files_for_later: Mutex::new(files_for_later),
+                    }),
+                    true,
+                ));
```
[the remark this answers is not in the document]
Comment (Joe-Blount) [reply compactor/src/components/round_info_source/mod.rs:695]: wow, yeah.  :laughing: I was moving that logic around while working on this PR.

File: compactor/src/driver.rs [hunk 12]
```diff
@@ -259,49 +259,90 @@ async fn try_compact_partition(
...
+            let branches_output: Vec<ParquetFile> = branches_output.into_iter().flatten().collect();
+
+            let mut files_for_later = range
+                .files_for_later
+                .lock()
+                .unwrap()
+                .take()
+                .unwrap_or(Vec::new());
+
+            files_for_later.extend(branches_output);
```
[the remark this answers is not in the document]
Comment (Joe-Blount) [reply compactor/src/driver.rs:339]: yes, this `branches_output` can be adding L0s to `files_for_later`.

File: compactor/tests/layouts/single_timestamp.rs (hunk 9 above)
Comment (Joe-Blount) [reply compactor/tests/layouts/single_timestamp.rs:151]: yes, and that's generally true of this file of tests.  The filename suggests all the tests have a single timestamp, but they're actually single or very low time ranges.

File: compactor/src/components/round_info_source/mod.rs (hunk 6 above)
Comment (Joe-Blount) [reply compactor/src/components/round_info_source/mod.rs:517]: this is `files_for_later` on the `RoundInfo`, not on an individual `CompactRange`.  The `CompactRange` `files_for_later` can be anything.  But the `RoundInfo` `files_for_later` will only be L2s.

From the `RoundInfo` struct:
```
    /// files_for_later holds L2 files while the ranges have L0s.  Splitting into many ranges is necessary for the L0->L1
    /// compactions, to keep the problem size manageable.  It would be unfortunate (inefficient) to split L2 files to prevent
    /// them from spanning multiple CompactRanges.  So instead, the L2 files are held in files_for_later until the L0s are gone,
    /// at which point we'll have a single range, which gets the L2 files.
    /// Note that each CompactRange also has its own files_for_later, which are L0/L1 files within that range that aren't being
    /// compacted in the current round.
    ```

File: compactor/src/components/divide_initial/multiple_branches.rs (hunk 7 above)
[the remark this answers is not in the document]
Comment (wiedld) [reply compactor/src/components/divide_initial/multiple_branches.rs:54]: Those [IR plan split times](https://github.com/influxdata/influxdb_iox/blob/80f8b55baa2480d3692f0a7f5f344e205f6ec4a9/compactor/src/components/ir_planner/planner_v1.rs#L251) are about splitting an actual file into multiple parts, not about dividing a group of files to then compact into a single file (per group). We do use `split_times` generated earlier, for both purposes. I was trying to update our terminology to make that clear. 🤷🏼  😆

File: compactor/src/components/round_info_source/mod.rs (hunk 6 above)
Comment (wiedld) [reply compactor/src/components/round_info_source/mod.rs:517]: Can we maybe re-name this as `l2_files_for_later`?

File: compactor/src/driver.rs (hunk 12 above)
[the remark this answers is not in the document]
Comment (wiedld) [reply compactor/src/driver.rs:339]: Thank you for delineating btwn the round_info's versus compact_range's `files_for_later`. I'm hoping a rename for the round_info's `l2_files_for_later` will make it a bit clearer, please!  🙏🏼

File: compactor/src/components/round_info_source/mod.rs (hunk 2 above)
Comment (wiedld) [reply compactor/src/components/round_info_source/mod.rs]: Removing that shadowing made it more clear. TY.

File: compactor/src/driver.rs (hunk 12 above)
[the remark this answers is not in the document]
Comment (Joe-Blount) [reply compactor/src/driver.rs:339]: named it generically because I'm not sure if it will stay l2 only.  But we can rename it again.  I'll push a commit to rename it.

[the remark this answers is not in the document]
Comment (wiedld) [reply compactor/src/driver.rs:339]: TY!

**Review by wiedld — APPROVED** [review]
TY for the code cleanup changes!
